### ISSUE TITLE
Memory leak

### ISSUE THREAD

**Issue by paladin-t** [libsdl-org/SDL#6200]
3 piece of memory leaks detected on windows with SDL2-2.24.0.

If an error occurs, SDL allocates an object to store the error in `SDL_GetErrBuf` in "SDL_thread.c".
But when is it freed?

https://github.com/libsdl-org/SDL/blob/d4e0d27c1c21e364d21be3af681f02b05f362345/src/thread/SDL_thread.c#L227

**Comment by slouken** [comment]
It's freed in `SDL_TLSCleanup()`

**Comment by paladin-t** [comment]
Thanks for the reply. But I'm afraid memory leak does exist.

In "SDL_systls.c", function `SDL_SYS_SetTLSData`, the operation `TlsSetValue(thread_local_storage, data)` was executed in the main thread.

But function `SDL_TLSCleanup` in "SDL_thread.c" was called from **another two threads** (`RunThreadViaCreateThread`), it makes function `SDL_SYS_GetTLSData` operation `TlsGetValue(thread_local_storage)` **WAS NOT** executed on the main thread, therefore `storage = SDL_SYS_GetTLSData()` in `SDL_TLSCleanup` always gets `NULL`, and there's no chance to call its `destructor` (`SDL_FreeErrBuf`).

The error occurs during `SDL_Init` in the main thread, `SDL_FreeErrBuf` was never called. This bug could be reproduced with either v2.24.0 or the latest main branch, Win32, either linked with SDL2 as a lib or dll.

**Comment by slouken** [comment]
This is intentional. There are applications that initialize and quit SDL multiple times, so the main thread never cleans up TLS data unless the application explicitly calls that function to signal that they are never going to call into SDL again. In most cases the application is going away so it doesn't matter.

TL;DR, if you want to free that memory, you should call `SDL_TLSCleanup()` after a call to SDL_Quit().

**Comment by slouken** [comment]
I went ahead and called SDL_TLSCleanup() from SDL_Quit(). I'm sure there was a reason we didn't in the past, so we'll see what happens.

**Comment by slouken** [comment]
Yep, I remember last time I looked it was pretty obvious, but we've done a bunch of defensive coding work since then, and I think whatever it was has been taken care of.

I tested running testgamecontroller in a loop without exiting and once I initialized the static variables (unrelated to this change), everything was working well.

The only thing I noticed is that the TLS slots are ever increasing and when we allocate TLS storage we allocate to the maximum slot that's ever been created, but that's not unique to this use case.
